### PR TITLE
[PLUGIN-1735] Fix read timeout issue during GCS Bucket  copy/move actions for large files

### PR DESCRIPTION
**Pull request by dj-smart**
[JIRA
](https://cdap.atlassian.net/browse/PLUGIN-1735
)
## Introduction:
- While copying large files using the **GCS Copy** plugin, `Read Timed out` error was occuring as Storage Client was configured with the default timeout of 20 seconds: https://cloud.google.com/storage/docs/retry-strategy#client-libraries

## Fix
- A property **readTimeout** has been introduced for GCS Copy/Move Plugins that enables user to provide the timeout based on their requirements depending on the file size that is being transferred.
- The Retry Settings for the Storage Client are configured to use this property using [this method](https://cloud.google.com/java/docs/reference/google-cloud-core/latest/com.google.cloud.http.HttpTransportOptions.Builder#com_google_cloud_http_HttpTransportOptions_Builder_setReadTimeout_int_)

## Testing
- Tested the changes by running a pipeline with **GCS Copy** plugin in CDAP sandbox

### CODE REVIEW

File: widgets/GCSMove-action.json [hunk 1]
```diff
@@ -128,6 +128,16 @@
           }
         }
       ]
+    },
+    {
+      "label" : "Advanced",
+      "properties" : [
+        {
+          "name": "readTimeout",
+          "widget-type": "textbox",
```
Comment (itsankit-google) [inline widgets/GCSMove-action.json]: use widget type number, for example: https://github.com/data-integrations/google-cloud/blob/b4c7d043eb84b6611c2d70554233d35cab32ff70/widgets/GCSFile-batchsource.json#L122

This comment applies to whole PR.

**Review by itsankit-google — COMMENTED** [review]
Please add a validation check for negative integers as well.

Please also include markdown changes.

**Comment by dj-smart** [comment]
> Please add a validation check for negative integers as well.
> 
> Please also include markdown changes.

1. For negative integers, it uses the default readTimeout as mentioned here: https://cloud.google.com/java/docs/reference/google-cloud-core/latest/com.google.cloud.http.HttpTransportOptions.Builder#com_google_cloud_http_HttpTransportOptions_Builder_setReadTimeout_int_
2. Added the documentation.

File: widgets/GCSMove-action.json (hunk 1 above)
Comment (dj-smart) [reply widgets/GCSMove-action.json]: Addressed

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java [hunk 2]
```diff
@@ -128,6 +139,7 @@ public void validate(FailureCollector collector, Map<String, String> arguments)
     collector.getOrThrowException();
   }
 
+
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:154]: Nit: remove empty line.

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java [hunk 3]
```diff
@@ -74,15 +75,25 @@ public class SourceDestConfig extends GCPConfig {
     " at https://cloud.google.com/data-fusion/docs/how-to/customer-managed-encryption-keys")
   protected String cmekKey;
 
+  @Name(READ_TIMEOUT)
+  @Macro
+  @Nullable
+  @Description("Timeout in seconds to read data from an established HTTP connection (Default value is 20). " +
+    ("For performing copy/move operation on large files in GCS buckets, set a higher value. " +
+      "Set it to 0 for infinite(no limit)"))
+  protected String readTimeout;
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java]: If the widget type is number, we can directly read it as Long saves an extra type conversion step.

File: widgets/GCSMove-action.json [hunk 4]
```diff
@@ -128,6 +128,16 @@
...
         }
       ]
+    },
+    {
+      "label" : "Advanced",
+      "properties" : [
+        {
+          "name": "readTimeout",
+          "widget-type": "number",
+          "label" : "Read Timeout"
```
Comment (itsankit-google) [inline widgets/GCSMove-action.json]: Can also set default minimum and maximum in the widget. See sample size widget referenced in last comment.

**Review by itsankit-google — CHANGES_REQUESTED** [review]
Please also add a validation check if negative value is passed in the config.

**Comment by itsankit-google** [comment]
> > Please add a validation check for negative integers as well.
> > Please also include markdown changes.
> 
> 1. For negative integers, it uses the default readTimeout as mentioned here: https://cloud.google.com/java/docs/reference/google-cloud-core/latest/com.google.cloud.http.HttpTransportOptions.Builder#com_google_cloud_http_HttpTransportOptions_Builder_setReadTimeout_int_
> 2. Added the documentation.

I saw that but we could just avoid it by a validation check to imply what is intended.

**Comment by itsankit-google** [comment]
Please always add a build label to the PR so that we know build is passing successfully before reviewing, it saves time for reviewers.

**Comment by dj-smart** [comment]
> Please always add a build label to the PR so that we know build is passing successfully before reviewing, it saves time for reviewers.

Acknowledged

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java (hunk 2 above)
Comment (dj-smart) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:154]: ACK

File: widgets/GCSMove-action.json (hunk 4 above)
Comment (dj-smart) [reply widgets/GCSMove-action.json]: Set the minimum and default values

**Comment by dj-smart** [comment]
> > > Please add a validation check for negative integers as well.
> > > Please also include markdown changes.
> > 
> > 
> > 
> > 1. For negative integers, it uses the default readTimeout as mentioned here: https://cloud.google.com/java/docs/reference/google-cloud-core/latest/com.google.cloud.http.HttpTransportOptions.Builder#com_google_cloud_http_HttpTransportOptions_Builder_setReadTimeout_int_
> > 2. Added the documentation.
> 
> I saw that but we could just avoid it by a validation check to imply what is intended.

Added the validation

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java (hunk 3 above)
Comment (dj-smart) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java]: Set the type to Integer

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java [hunk 5]
```diff
@@ -125,9 +136,20 @@ public void validate(FailureCollector collector, Map<String, String> arguments)
...
     collector.getOrThrowException();
   }
-
+  void validateReadTimeout(FailureCollector collector) {
+    if (readTimeout == null) {
+      return;
+    }
+    if (readTimeout < 0) {
+      collector.addFailure("Please enter zero or a positive value"
+        , null).withConfigProperty(READ_TIMEOUT);
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java]:
```
collector.addFailure("Read Timeout cannot be less than 0", "Please enter 0 or positive value").withConfigProperty(READ_TIMEOUT);
```

```Please enter 0 or positive value``` is a corrective action.

**Review by itsankit-google — COMMENTED** [review]
one minor comment, please squash commits using rebase and test locally.

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java (hunk 5 above)
Comment (dj-smart) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java]: Understood. Done.

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java [hunk 6]
```diff
@@ -125,9 +136,20 @@ public void validate(FailureCollector collector, Map<String, String> arguments)
     if (!containsMacro(NAME_CMEK_KEY)) {
       validateCmekKey(collector, arguments);
     }
+    if (!containsMacro(READ_TIMEOUT)) {
+      validateReadTimeout(collector);
+    }
     collector.getOrThrowException();
   }
-
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:130]: why is this removed? There should be an empty line between two methods.

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java [hunk 7]
```diff
@@ -125,9 +136,20 @@ public void validate(FailureCollector collector, Map<String, String> arguments)
...
+  void validateReadTimeout(FailureCollector collector) {
+    if (readTimeout == null) {
+      return;
+    }
+    if (readTimeout < 0) {
+      collector.addFailure("Read Timeout cannot be less than 0",
+                           "Please enter 0 or a positive value").withConfigProperty(READ_TIMEOUT);
+    }
+  }
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:153]: similarly here, add an empty line before another method starts.

Comment (dj-smart) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:153]: Done

File: src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java (hunk 6 above)
Comment (dj-smart) [reply src/main/java/io/cdap/plugin/gcp/gcs/actions/SourceDestConfig.java:130]: Done

File: src/main/java/io/cdap/plugin/gcp/gcs/StorageClient.java [hunk 8]
```diff
@@ -320,17 +321,21 @@ private static String toPath(BlobId blobId) {
...
+      builder.setTransportOptions(HttpTransportOptions.newBuilder().setReadTimeout(readTimeout * 1000).build());
+    }
     Storage storage = builder.build().getService();
     return new StorageClient(storage);
   }
 
-  public static StorageClient create(GCPConnectorConfig config) throws IOException {
-    return create(config.getProject(), config.getServiceAccount(), config.isServiceAccountFilePath());
+  public static StorageClient create(GCPConnectorConfig config, Integer readTimeout) throws IOException {
+    return create(config.getProject(), config.getServiceAccount(), config.isServiceAccountFilePath(), readTimeout);
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/gcp/gcs/StorageClient.java]:
```
public static StorageClient create(GCPConnectorConfig config) throws IOException {
    return create(config.getProject(), config.getServiceAccount(), config.isServiceAccountFilePath(),  null);
```
This way we will not need to change the underlying classes which are calling it and not using `readTimeout`.

Comment (dj-smart) [reply src/main/java/io/cdap/plugin/gcp/gcs/StorageClient.java]: Just to clarify, are you proposing to add this additional method along with the existing one?

Comment (dj-smart) [reply src/main/java/io/cdap/plugin/gcp/gcs/StorageClient.java]: Discussed offline, made the change

**Review by itsankit-google — APPROVED** [review]
